drivers: sensor: grow_r502a: Add Missing functionalities and improvements for GROW_R502A fingerprint sensor

DineshDK03 opened the pull request:
Add the following functionality and improvements for the R502A fingerprint sensor device.

1. Improvements and bug fixes for the existing mainline version.
2. Add capture, create, load, store, and compare attributes to work with template data from the finger or from the sensor's library index.
3. Add LED control for the R502A sensor's built-in LED.
4. Add upload and download of fingerprint template data.
5. Add Read and write system parameters of the R502A sensor.

DineshDK03 committed:
drivers: sensor: grow_r502a: fix in fps_search

return -ENOENT; when no record found on getting
attribute SENSOR_ATTR_R502A_FIND and fix array indexing
for tx data in fps_search.

Signed-off-by: Dinesh Kumar K <dinesh@linumiz.com>

DineshDK03 committed:
drivers: sensor: grow_r502a: modified Tx and Rx data handling

improvised uart tx and rx handler functions to handle
large chunk of fingerprint data from/to the sensor.

Signed-off-by: Dinesh Kumar K <dinesh@linumiz.com>

DineshDK03 committed:
drivers: sensor: grow_r502a: add built-in LED control

Add LED control attribute for sensor device's
built-in LED using zephyr's led driver APIs.

Signed-off-by: Dinesh Kumar K <dinesh@linumiz.com>

DineshDK03 committed:
drivers: sensor: grow_r502a: add set/read system parameter

Add functionality for read and set system
parameter of sensor device.

Signed-off-by: Dinesh Kumar K <dinesh@linumiz.com>

DineshDK03 committed:
drivers: sensor: grow_r502a: add upload and download template

1. Add functionality for upload and download fingerprint
template data of the R502A fingerprint sensor device.
2. change data type for data_len parameter of
transceive_packet function to uint16_t as it may hold
value 256 for dowload template data to sensor.

Signed-off-by: Dinesh Kumar K <dinesh@linumiz.com>

DineshDK03 committed:
drivers: sensor: grow_r502a: add RX packet validation

Add RX packet validator where header, address, pid
and checksum bytes are verified and validated on
every functions.

Signed-off-by: Dinesh Kumar K <dinesh@linumiz.com>

DineshDK03 committed:
drivers: sensor: grow_r502a: move enroll, search, delete to application

Removed fps_enroll, fps_search and fps_delete from
driver code, since they did application usage inside
driver. add the following attributes to achieve the
above functionality from application.
1. SENSOR_ATTR_R502A_CAPTURE
2. SENSOR_ATTR_R502A_TEMPLATE_CREATE
3. SENSOR_ATTR_R502A_RECORD_ADD
4. SENSOR_ATTR_R502A_RECORD_LOAD
5. SENSOR_ATTR_R502A_COMPARE

Signed-off-by: Dinesh Kumar K <dinesh@linumiz.com>

MaureenHelm commented on drivers/sensor/grow_r502a/grow_r502a.c:
```diff
@@ -287,6 +288,7 @@ static int fps_image_to_char(const struct device *dev, uint8_t char_buf_idx)
 	struct grow_r502a_data *drv_data = dev->data;
 	union r502a_packet rx_packet = {0};
 	char const img_to_char_len = 2;
+	int ret = 0;
```
Are the `ret` changes in this commit related to `drivers: sensor: grow_r502a: move enroll, search, delete to application`? It looks like they may belong in a separate commit

MaureenHelm commented on drivers/sensor/grow_r502a/grow_r502a.c:
```diff
@@ -461,8 +461,8 @@ static int fps_search(const struct device *dev, uint8_t char_buf_idx)
-	sys_put_be16(R02A_LIBRARY_START_IDX, &tx_packet.data[1]);
-	sys_put_be16(R502A_DEFAULT_CAPACITY, &tx_packet.data[3]);
+	sys_put_be16(R02A_LIBRARY_START_IDX, &tx_packet.data[2]);
+	sys_put_be16(R502A_DEFAULT_CAPACITY, &tx_packet.data[4]);
```
`drivers: sensor: grow_r502a: improvements and bug fixes` should be split up into separate commits for each logical bug fix

MaureenHelm commented on include/zephyr/drivers/sensor/grow_r502a.h:
```diff
@@ -71,6 +124,12 @@ enum sensor_attribute_grow_r502a {
+	 * @param val->val1	use R502A_LED_PARAMS macro with desired
+	 *			LED configurations.
+	 */
+	SENSOR_ATTR_R502A_DEVICE_LED,
```
Please look at enabling the LED driver API

DineshDK03 replied:
LED is internal to the Fingerprint sensor and is accessible to indicate fps actions alone. I don't think we need this LED-controlling functionality to be part of the LED driver. please help me understand why and how? 
  
![image](https://github.com/zephyrproject-rtos/zephyr/assets/37179801/1742cb59-aabc-418f-8a1e-db1460ac94e0)

DineshDK03 replied:
@MaureenHelm ping

DineshDK03 replied:
@MaureenHelm ping

MaureenHelm replied:
It looks like you're enabling functionality similar to the LED API (on, off, blink, breathe). Rather than do it through a custom sensor attribute, I'm suggesting that you implement the LED API for this device

DineshDK03 replied:
@MaureenHelm Since this grow_r502a Fingerprint sensor comes under the sensor subsystem and LED is one of the attributes of this grow_r502a sensor, there won't be any need to control this LED as a separate entity, I think it would be best to control this FPS LED with SENSOR_ATTR alone.
If we implement LED API, does it not create any conflict of one device 2 driver API set?

DineshDK03 replied:
@MaureenHelm ping

DineshDK03 replied:
> @MaureenHelm Since this grow_r502a Fingerprint sensor comes under the sensor subsystem and LED is one of the attributes of this grow_r502a sensor, there won't be any need to control this LED as a separate entity, I think it would be best to control this FPS LED with SENSOR_ATTR alone. If we implement LED API, does it not create any conflict of one device 2 driver API set?

@MaureenHelm ping

teburd replied:
This can be somewhat resolved by having a single device's data/config structs be shared between two struct device's implementing the two APIs perhaps... but right its a bit confusing how this is actually supposed to work I agree. I think @bjarki-trackunit implemented multiple device APIs for a device before maybe he has some comments on this.

DineshDK03 commented on include/zephyr/drivers/sensor.h:
```diff
@@ -50,6 +63,8 @@ struct sensor_value {
 	int32_t val1;
 	/** Fractional part of the value (in one-millionth parts). */
 	int32_t val2;
+	/** buffer part of the value*/
+	struct sensor_value_ex ex;
```
It is important for larger chunks of data exported or imported to the driver and introducing ex is inevitable, the footprint is just 8 bytes in a 32-bit machine or shall be made as a confidential compilation.

DineshDK03 replied:
@MaureenHelm ping

teburd replied:
Unfortunately I agree with Maureen in this case. I don't think this change can be accepted.

Could the sensor_value two words be used to store the buf pointer and length instead?

Something like...

``` c
struct sensor_value fp_data = {
   .val1 = (int32_t)(buf),
   .val2 = (int32_t)(buf_len),
}
```

I'd note though that passing buffers to/from user mode threads will not work without special care to share the memory from kernel space to user space. So that would be a caveat for this particular driver and use case.

yperess replied:
I would recommend using the async API instead.

teburd replied:
There's an attribute that this seems to be used for

DineshDK03 committed:
drivers: sensor: grow_r502a: fix in fps_empty_db

return ret; on goto unlock in fps_empty_db function.

Signed-off-by: Dinesh Kumar K <dinesh@linumiz.com>

DineshDK03 committed:
drivers: sensor: grow_r502a: fix in checksum calculation

change checksum calculation to consider 2 bytes in pkg_len.

Signed-off-by: Dinesh Kumar K <dinesh@linumiz.com>

DineshDK03 committed:
drivers: sensor: grow_r502a: fix in fps_read_template_table

consuming sensor_value struct instead of local device
driver variable.

Signed-off-by: Dinesh Kumar K <dinesh@linumiz.com>

DineshDK03 commented:
@MaureenHelm please have a look at the comments and do a re-review if needed.

DineshDK03 commented:
@MaureenHelm Please revisit

teburd reviewed (changes requested):
The sample should be a fingerprint general sample IMO, if generic LED control is desired alongside it DT can be used to connect a fingerprint sensor and a RGB LED

Making them two distinct devices would enable a lot more combinations of this sort of thing than having it all in one sensor device driver.

yperess reviewed (changes requested):
My suggestion in order to get this PR merged is that since the new data representation is needed for a custom attribute that's declared in the driver's `.h` file, why not just have that attribute be settable via a custom function? It's really no different. So basically users of this driver instead of calling:

```c
sensor_attr_set(
    dev,
    SENSOR_CHAN_FINGERPRINT,
    SENSOR_ATTR_R502A_DOWNLOAD,
    &value
);
```

They would use a specific function like:
```c
grow_r502a_download(dev, buffer, buffer_len);
```

teburd commented:
sensor wg:
* fingerprint related calls should be driver specific (expose r502a specific functions) as needed
* led if states are internally defined then its ok, if controllable using the API then should be a seperate struct device

DineshDK03 commented:
updated the PR as per the discussions https://discord.com/channels/720317445772017664/1163527713290719373  please have a look and review guys @teburd @yperess

teburd reviewed (dismissed):
I'm ok with this approach, needs to pass CI though

kartben commented on samples/sensor/grow_r502a/src/main.c:
```diff
@@ -64,13 +119,29 @@ static void trigger_handler(const struct device *dev,
+		return;
+	}
+
+	printk("baud %d\n", res.baud);
+	printk("addr 0x%x\n", res.addr);
+	printk("lib_size %d\n", res.lib_size);
+	printk("data_pkt_size %d\n", res.data_pkt_size);
```
not really blocking but I think it would be great to update the contents of https://docs.zephyrproject.org/latest/samples/sensor/grow_r502a/README.html  so that these changes are reflected in the output of the program (there might also be other verbiage worth updating in this page?)

DineshDK03 replied:
Updated. Thanks for this comment.

DineshDK03 committed:
samples: sensor: Modified Readme for grow_r502a

Modified readme for grow_r502a sample w.r.t
latest changes on driver like attributes and
LED controlling via zephyr LED subsystem APIs.

Signed-off-by: Dinesh Kumar K <dinesh@linumiz.com>
